WIP: getting started on supporting `nonlocal` statements

oconnor663 opened the pull request:
Putting up a draft PR so I can ask questions about it. Comments below.

oconnor663 commented on crates/ty_python_semantic/resources/mdtest/scopes/global.md:
```diff
@@ -83,7 +83,7 @@ def f():
     def g() -> None:
         nonlocal x
-        global x  # TODO: error: [invalid-syntax] "name 'x' is nonlocal and global"
+        global x  # error: [invalid-syntax] "name 'x' is nonlocal and global"
```
This `error:` assert should start passing when I fill in some other TODOs below.

oconnor663 commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
````diff
@@ -43,3 +43,64 @@ def f():
             reveal_type(x)  # revealed: Unknown | Literal[1]
 ```
+
+## The `nonlocal` keyword
````
This new test is currently passing. Yay.

oconnor663 commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
````diff
@@ -43,3 +43,64 @@ def f():
+        x += 1
+```
+
+## TODO: get the write to land in the right scope
````
From here on down, it's not currently passing. These are the bits I need help with.

oconnor663 commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1875,11 +1879,50 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
+                    if let Some(nonlocals) = self.nonlocals_by_scope.get(&scope_id)
+                        && nonlocals.contains(&symbol_id)
+                    {
+                        // TODO: This should be some sort of error?
```
I'm not sure how to report an error/diagnostic here, but once I figure that out (the same for its partner in the `Stmt::Nonlocal` branch immediately below) that should make the `error:` assert in `global.md` start passing.

oconnor663 replied:
e3fc88b94af80f5a93157384508c5ec06376c945

oconnor663 commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1875,11 +1879,50 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
                         .insert(symbol_id);
                 }
+                // TODO: why is this here?
+                walk_stmt(self, stmt);
```
It seems like this does nothing? Curious what's up here.

AlexWaygood replied:
sort of a similar principle in spirit to [flake8-bugbear](https://github.com/PyCQA/flake8-bugbear)'s B906 rule (not yet reimplemented in Ruff, alas)

oconnor663 commented on crates/ty_python_semantic/src/types/infer.rs:
```diff
@@ -5692,14 +5692,16 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
-                if skip_non_global_scopes {
+            let mut is_nonlocal_binding = false;
+            if let Some(name) = expr.as_name()
+                && let Some(symbol_id) = place_table.place_id_by_name(name)
```
Probably we don't want to start using let-chains immediately and bump the MSRV to 1.88? But I figured I'd ask :)

carljm reviewed (commented):
Great start! Tbh I think it would be fine to land something close to this (maybe get the easy syntax errors working) with TODOs for the impact on revealed type in outer scope, and handle that in separate follow-up.

oconnor663 commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1875,11 +1879,50 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
+                    let symbol = symbol_table.place_expr(symbol_id);
+                    if symbol.is_bound() || symbol.is_declared() || symbol.is_used() {
+                        self.report_semantic_error(SemanticSyntaxError {
+                            kind: SemanticSyntaxErrorKind::LoadBeforeGlobalDeclaration {
```
e3fc88b94af80f5a93157384508c5ec06376c945 fixes this and adds one test. I see that the [`Semantic syntax errors` section of `globals.md`](https://github.com/astral-sh/ruff/blob/77a5c5ac80cfe77def2397e80ed9a7d79e068b34/crates/ty_python_semantic/resources/mdtest/scopes/global.md#semantic-syntax-errors) tests like a dozen different variants of this. Should I maybe copy all of those here?

oconnor663 commented on crates/ty_python_semantic/resources/mdtest/scopes/global.md:
```diff
@@ -211,3 +211,16 @@ def f():
+def f():
+    global x
+
+# TODO: reveal_type(x)  # revealed: Unknown | Literal["1"]
```
Should I open an issue for this TODO?

AlexWaygood replied:
Deferring it for now and opening a followup sounds great to me!

oconnor663 commented:
@AlexWaygood this is now plausibly ready for a proper review :) I did a _lot_ of copy-pasting here, from existing `global` code in particular and from surrounding code in general, and I expect there to be a lot of cases where I'm round-tripping between e.g. `Scope` and `FileScopeId` way more than necessary, or generally doing things in un-idiomatic ways.

oconnor663 commented on crates/ruff_linter/src/checkers/ast/mod.rs:
```diff
@@ -671,7 +671,11 @@ impl SemanticSyntaxContext for Checker<'_> {
+            | SemanticSyntaxErrorKind::LoadBeforeNonlocalDeclaration { .. }
+            | SemanticSyntaxErrorKind::NonlocalAndGlobal(_)
+            | SemanticSyntaxErrorKind::AnnotatedGlobal(_)
+            | SemanticSyntaxErrorKind::AnnotatedNonlocal(_) => {
```
@ntBre did you mention that you'd be willing to help me fill in "proper" Ruff support for these new cases? :)

ntBre replied:
Yeah, for sure! I'll either handle them myself or at least add them to [#17412](https://github.com/astral-sh/ruff/issues/17412).

oconnor663 commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1422,6 +1426,35 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
                     self.visit_expr(value);
                 }
 
+                if let ast::Expr::Name(name) = &*node.target {
+                    let symbol_id = self.add_symbol(name.id.clone());
+                    let scope_id = self.current_scope();
+                    // Check whether the variable has been declared global.
+                    if let Some(globals) = self.globals_by_scope.get(&scope_id) {
+                        if globals.contains(&symbol_id) {
+                            self.report_semantic_error(SemanticSyntaxError {
+                                kind: SemanticSyntaxErrorKind::AnnotatedGlobal(
+                                    name.id.as_str().into(),
+                                ),
+                                range: name.range,
+                                python_version: self.python_version,
+                            });
+                        }
+                    }
+                    // Check whether the variable has been declared nonlocal.
+                    if let Some(nonlocals) = self.nonlocals_by_scope.get(&scope_id) {
+                        if nonlocals.contains(&symbol_id) {
+                            self.report_semantic_error(SemanticSyntaxError {
+                                kind: SemanticSyntaxErrorKind::AnnotatedNonlocal(
+                                    name.id.as_str().into(),
+                                ),
+                                range: name.range,
+                                python_version: self.python_version,
+                            });
+                        }
+                    }
+                }
+
```
Hah no of course there's no reason, I just haven't noticed `SemanticSyntaxChecker` yet :)

AlexWaygood commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
```diff
@@ -43,3 +43,272 @@ def f():
+
+## The `nonlocal` keyword
+
+Without the `nonlocal` keyword, `x += 1` is not allowed in an inner scope, even if we break it up
```
"`x += 1` is not allowed in an inner scope" feels a bit confusing to me here -- this would obviously be allowed, because here `x` is unambiguously a local variable within the scope of `g`:

```py
def f():
    x = 1
		def g():
		    x = 1
		    x += 1
```

oconnor663 replied:
d473f0e1dcb27d3b4cee0f648b5cde241c762819

AlexWaygood commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
````diff
@@ -43,3 +43,272 @@ def f():
+        x = y + 1
+```
+
+With `nonlocal` these examples work:
````
```suggestion
With `nonlocal` these examples work, because the reassignments modify the variable from the outer scope rather than modifying a separate variable local to the inner scope.
```

AlexWaygood commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
````diff
@@ -43,3 +43,272 @@ def f():
+        nonlocal x, y  # error: [invalid-nonlocal] "no binding for nonlocal `y` found"
+```
+
+A global `x` doesn't work either; it has to be a function-like scope:
````
```suggestion
A global `x` doesn't work either; the outer-scope binding for the variable has to originate from a function-like scope:
```

AlexWaygood commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
these tests are fantastically thorough -- thank you!

UnboundVariable replied:
Yeah, nice tests!

AlexWaygood reviewed (commented):
A review of the tests, which are overall fantastic!

AlexWaygood commented on crates/ty_python_semantic/src/types/infer.rs:
```diff
@@ -1564,6 +1565,7 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         let mut bound_ty = ty;
 
         let global_use_def_map = self.index.use_def_map(FileScopeId::global());
+        let nonlocal_use_def_map;
```
I had no idea this was legal in Rust, so thanks for teaching me this new trick 😃 However, I do think I'd find the code slightly easier to understand if we used a `for` loop below rather than `loop` -- maybe something like this? (All your tests pass with this patch applied)

```diff
diff --git a/crates/ty_python_semantic/src/types/infer.rs b/crates/ty_python_semantic/src/types/infer.rs
index 9f80d70b7b..9860f1d8ec 100644
--- a/crates/ty_python_semantic/src/types/infer.rs
+++ b/crates/ty_python_semantic/src/types/infer.rs
@@ -1565,7 +1565,7 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         let mut bound_ty = ty;
 
         let global_use_def_map = self.index.use_def_map(FileScopeId::global());
-        let nonlocal_use_def_map;
+        let mut nonlocal_use_def_map = None;
         let place_id = binding.place(self.db());
         let place = place_table.place_expr(place_id);
         let skip_non_global_scopes = self.skip_non_global_scopes(file_scope_id, place_id);
@@ -1585,47 +1585,51 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
             .symbol_is_nonlocal_in_scope(place_id, file_scope_id)
         {
             // Walk up parent scopes looking for the enclosing scope that has definition of this
-            // name. `ancestor_scopes` includes the current scope, so skip that one.
-            let mut ancestor_scopes = self.index.ancestor_scopes(file_scope_id).skip(1);
-            loop {
-                if let Some((enclosing_scope_file_id, enclosing_scope)) = ancestor_scopes.next() {
-                    // Ignore class scopes and the global scope.
-                    if !enclosing_scope.kind().is_function_like() {
-                        continue;
-                    }
-                    let enclosing_place_table = self.index.place_table(enclosing_scope_file_id);
-                    let Some(enclosing_place_id) =
-                        enclosing_place_table.place_id_by_expr(&place.expr)
-                    else {
-                        // This ancestor scope doesn't have a binding. Keep going.
-                        continue;
-                    };
-                    if self
-                        .index
-                        .symbol_is_nonlocal_in_scope(enclosing_place_id, enclosing_scope_file_id)
-                    {
-                        // The variable is `nonlocal` in this ancestor scope. Keep going.
-                        continue;
-                    }
-                    if self
-                        .index
-                        .symbol_is_global_in_scope(enclosing_place_id, enclosing_scope_file_id)
-                    {
-                        // The variable is `global` in this ancestor scope. This "breaks" the
-                        // `nonlocal` "chain", and it's a syntax error in
-                        // `infer_nonlocal_statement`. Ignore it here and bail out of this loop.
-                        break use_def.declarations_at_binding(binding);
-                    }
-                    // We found the closest definition. Note that (unlike in `infer_place_load`)
-                    // this does *not* need to be a binding. It could be just `x: int`.
-                    nonlocal_use_def_map = self.index.use_def_map(enclosing_scope_file_id);
-                    break nonlocal_use_def_map.end_of_scope_declarations(enclosing_place_id);
+            // name. If one is found, it will be saved as the `nonlocal_use_def_map` variable.
+            //
+            //`ancestor_scopes` includes the current scope, so skip that one.
+            for (enclosing_scope_file_id, enclosing_scope) in
+                self.index.ancestor_scopes(file_scope_id).skip(1)
+            {
+                // Ignore class scopes and the global scope.
+                if !enclosing_scope.kind().is_function_like() {
+                    continue;
                 }
-
-                // We ran out of ancestor scopes without finding a definition. This case is a
-                // syntax error (no binding for `nonlocal` found) but ignore that here.
-                break use_def.declarations_at_binding(binding);
+                let enclosing_place_table = self.index.place_table(enclosing_scope_file_id);
+                let Some(enclosing_place_id) = enclosing_place_table.place_id_by_expr(&place.expr)
+                else {
+                    // This ancestor scope doesn't have a binding. Keep going.
+                    continue;
+                };
+                if self
+                    .index
+                    .symbol_is_nonlocal_in_scope(enclosing_place_id, enclosing_scope_file_id)
+                {
+                    // The variable is `nonlocal` in this ancestor scope. Keep going.
+                    continue;
+                }
+                if self
+                    .index
+                    .symbol_is_global_in_scope(enclosing_place_id, enclosing_scope_file_id)
+                {
+                    // The variable is `global` in this ancestor scope. This "breaks" the
+                    // `nonlocal` "chain", and it's a syntax error in
+                    // `infer_nonlocal_statement`. Ignore it here and bail out of this loop.
+                    break;
+                }
+                // We found the closest definition. Note that (unlike in `infer_place_load`)
+                // this does *not* need to be a binding. It could be just `x: int`.
+                nonlocal_use_def_map = Some((
+                    self.index.use_def_map(enclosing_scope_file_id),
+                    enclosing_place_id,
+                ));
+                break;
             }
+
+            nonlocal_use_def_map
+                .as_ref()
+                .map(|(map, place_id)| map.end_of_scope_declarations(*place_id))
+                .unwrap_or_else(|| use_def.declarations_at_binding(binding))
         } else {
             use_def.declarations_at_binding(binding)
         };
```

oconnor663 replied:
b31102a9eea7e52b19e0bce3646f8d53510add9c

AlexWaygood reviewed (commented):
The code looks fantastic. I agree with @MichaReiser's comments that it feels like some of the logic in `infer_nonlocal_statement` might ideally go in the `SemanticSyntaxChecker` rather than ty, though -- then Ruff and ty can naturally share the same code rather than us having to implement it twice for the two static-analysis tools 😃 if that would be hard to do, though, I also think it would be fine to land an initial implementation here and then work towards porting it to the `SemanticSyntaxChecker` (or adding a standalone separate implementation in Ruff!) as a followup

MichaReiser commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -104,6 +104,7 @@ pub(super) struct SemanticIndexBuilder<'db, 'ast> {
     scopes_by_node: FxHashMap<NodeWithScopeKey, FileScopeId>,
     scopes_by_expression: FxHashMap<ExpressionNodeKey, FileScopeId>,
     globals_by_scope: FxHashMap<FileScopeId, FxHashSet<ScopedPlaceId>>,
+    nonlocals_by_scope: FxHashMap<FileScopeId, FxHashSet<ScopedPlaceId>>,
```
I considered something similar but decided not to comment because I considered it out of scope for this PR. Ideally, we'd move all those rarely used fields (includes generators) into an `Option<Box<ExtraFields>>` where `ExtraFields` is only allocated when the file makes use of the extra field. 

But I then came to the conclusion that the overhead is probably minimal, considering that we only create a semantic index per file (and not per scope or even more granular). That's why I'm not sure this optimization matters and we should go with what's easier.

AlexWaygood replied:
yeah, I definitely know what you mean. I think my proposed refactor would probably not _significantly_ complicate things and I do also like that it has the nice property that it more closely models the runtime semantics (a symbol can only ever be `global`, `nonlocal` or neither; it can never be both `global` _and_ `nonlocal`). But you're right that this refactor definitely wouldn't be worth a big increase in complexity.

oconnor663 replied:
Not sure whether this is already clear, but as far as I know most Rust hash maps don't actually allocate when they're empty. So if the main performance concern is heap allocation, and these containers are usually empty, we might not need to worry about it?

MichaReiser replied:
There's a NIT concern that each hash map (even empty once) take 48 bytes  (I think) on 64bit systems of storage in its enclosing struct. I don't think this is worth a long discussion and @oconnor663 should just pick one approach.

oconnor663 replied:
I noticed that we already have `MARKED_GLOBAL` and `MARKED_NONLOCAL` defined in `PlaceFlags`, and we just weren't using them yet. I think it's cleaner to delete the maps entirely and use the flags: 27043edabdfcfb794896bf88ab785cb7513b4139

oconnor663 commented on crates/ty_python_semantic/src/types/diagnostic.rs:
```diff
@@ -1560,6 +1561,25 @@ declare_lint! {
+    /// Detects `nonlocal` statements that don't match a binding in any enclosing scope
+    ///
+    /// ## Why is this bad?
+    /// Unmatched `nonlocal` statements will raise a `SyntaxError` at runtime.
```
I think this was originally the just first place I tried adding new error types, and eventually I moved all the others to `SemanticSyntaxErrorKind`, and probably there's no longer any good reason for this to be here :sweat_smile:

oconnor663 commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
```diff
@@ -43,3 +43,272 @@ def f():
+## `nonlocal` bindings respect declared types from the defining scope
+
+By default (without `nonlocal`), an inner variable shadows an outer variable of the same name, and
+type declarations from outer scope don't apply to the inner one:
```
04efabb84

MichaReiser commented:
I'll put this back in Draft. You can republish it when it is ready for another review. This makes it easier for reviewers to know when it's ready for another review.

oconnor663 commented:
This PR is ready for another round of review. The big remaining question is whether to move the most complicated `nonlocal` check out of `infer.rs` and into `SemanticSyntaxError`. Basically this example:
```py
def f():
    def g():
        def h():
            # This is allowed, because of the subsequent definition of `x` two scopes up.
            nonlocal x
    x = 1
```
As it stands, I still have this check in `infer.rs`, which requires(?) declaring `INVALID_NONLOCAL` with `declare_lint!`, which we don't love. Two thoughts about that:
1. The best way to implement this check is probably to "defer" walking function bodies until after we've walked all the statements in enclosing scopes. This is what Ruff does today. This would still be one walk of the AST, but in the example above by the time we encountered `nonlocal x` we'd already have seen `x = 1`, and we could check the `nonlocal` requirement immediately. @erictraut [mentioned in chat](https://discord.com/channels/1039017663004942429/1343690517745111081/1392652928233508977) that that's how Pyright does it. However, even if we want to do that refactoring (any objections?), I'd rather not do it in this PR.
2. Even with that refactoring, it's not obvious to me how to move the check into `SemanticSyntaxChecker`. I see three options, but all of them look problematic:
  - If we added a simple boolean function like e.g. `is_nonlocal_symbol_available(name)` to the `SemanticSyntaxContext` trait, then Ruff and Ty would need to duplicate the scope walking, `global`-breaks-the-chain logic (arguably the part we'd most like to share?) to implement that function.
  - We could instead keep track of entering and exiting scopes in `SemanticSyntaxChecker`. It looks like we used to do this, but [we removed it back in April](https://github.com/astral-sh/ruff/pull/17314). I haven't read all the context there, but it makes me question adding it back.
  - We could add methods to the `SemanticSyntaxContext` trait that let the checker query scopes and bindings. This seems tricky because the model types are totally different(?) on the Ruff and Ty sides, so we'd be looking at adding a third(?) set of ID types and converting back and forth? Existing functions like `.global(name: &str)` have avoided doing that by leaving the "current scope" implicit, but that won't be enough for this check.

I think regardless of which of those three options we prefer, I'd rather leave it TODO in this PR and open a followup ticket. Let me know what you guys think?

UnboundVariable commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
````diff
@@ -43,3 +43,295 @@ def f():
+                nonlocal x, y, z  # error: [invalid-nonlocal] "no binding for nonlocal `y` found"
+                x = "string"  # error: [invalid-assignment]
+                z = "string"  # not an error
+```
````
Here's another combination of nested functions and classes with type declarations.

```python
def f():
    x: int = 1

    class Foo:
        x: str = ""

        @staticmethod
        def g():
            nonlocal x
            x = 1  # not an error
            x = ""  # error: [invalid-assignment]
```

oconnor663 replied:
also 3aaed2385744667b3d188b1a9084802095d163f9

MichaReiser reviewed (approved):
Nice job with removing the extra hash map! 

This is good to go once we removed the new lint (see my inline comment on how to accomplish that). 

> The best way to implement this check is probably to "defer" walking function bodies until after we've walked all the statements in enclosing scopes. This is what Ruff does today. This would still be one walk of the AST, but in the example above by the time we encountered nonlocal x we'd already have seen x = 1, and we could check the nonlocal requirement immediately. @erictraut [mentioned in chat](https://discord.com/channels/1039017663004942429/1343690517745111081/1392652928233508977) that that's how Pyright does it. However, even if we want to do that refactoring (any objections?), I'd rather not do it in this PR.

I think this is worth a try (scoped to a few hours). I also think it's fine to defer integrating this into the `SemanticSyntaxChecker` for now (maybe something @ntBre could help with as well). Would you mind opening an issue and adding TODO comments in the code (above the blocks that should be moved).

AlexWaygood commented on crates/ty_python_semantic/src/semantic_index/place.rs:
```diff
@@ -816,6 +824,14 @@ impl PlaceTableBuilder {
         self.table.places[id].insert_flags(PlaceFlags::IS_USED);
     }
 
+    pub(super) fn mark_place_global(&mut self, id: ScopedPlaceId) {
+        self.table.places[id].insert_flags(PlaceFlags::MARKED_GLOBAL);
+    }
+
+    pub(super) fn mark_place_nonlocal(&mut self, id: ScopedPlaceId) {
+        self.table.places[id].insert_flags(PlaceFlags::MARKED_NONLOCAL);
+    }
```
oh, this is a much better approach. Nice work!

AlexWaygood commented on crates/ty_python_semantic/src/types/infer.rs:
```diff
@@ -4611,6 +4662,71 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
+                builder
+                    .into_diagnostic(format_args!("no binding for nonlocal `{name}` found"))
+                    .annotate(Annotation::primary(
+                        Span::from(self.context.file()).with_range(*range),
```
```suggestion
                        self.context.span(*range),
```

MichaReiser replied:
TIL

AlexWaygood reviewed (approved):
Great work!

oconnor663 commented on crates/ty_python_semantic/resources/mdtest/scopes/nonlocal.md:
````diff
@@ -43,3 +43,295 @@ def f():
+        nonlocal x  # error: [invalid-nonlocal] "no binding for nonlocal `x` found"
+```
+
+A class-scoped `x` also doesn't work:
````
Thanks! 3aaed2385744667b3d188b1a9084802095d163f9

oconnor663 commented:
> I think this is worth a try (scoped to a few hours). I also think it's fine to defer integrating this into the `SemanticSyntaxChecker` for now (maybe something @ntBre could help with as well). Would you mind opening an issue and adding TODO comments in the code (above the blocks that should be moved).

https://github.com/astral-sh/ruff/pull/19271 is a draft PR that tries to do this. It actually seems to work for the `nonlocal.md` tests, but it causes a lot of seemingly unrelated failures that I'm having trouble debugging. Would anyone have time to take a look at it (details in the PR description)?

carljm commented:
Haven't had a chance to look at the PR, but I expect that switching to a breadth first AST walk in semantic indexing would require significant reworking of our handling of eager nested scopes (finding the right bindings in an outer scope at the point where the eager nested scope is defined).

AlexWaygood commented:
My inclination would just be to land this as-is for now, and punt on questions such as how we implement this syntax error in Ruff, whether we should try to integrate this syntax error into the `SemanticSyntaxChecker`, and whether we should attempt a significant refactor of the `SemanticIndexBuilder`. This PR gets the Python semantics correct, I don't think it matters _too_ much in which file we implement the syntax error detection, and landing it in its current state doesn't preclude us making any of those other improvements at a later date!
